File: src/Algorithm/LeetCode/LeetCode_337_House_Robber_III/Solution_DP.java
package Algorithm.LeetCode.LeetCode_337_House_Robber_III;

//https://www.programmercarl.com/0337.打家劫舍III.html
public class Solution_DP {
    public int rob(TreeNode root) {
        int[] res = robAction(root);

        return Math.max(res[0], res[1]);
    }

    private int[] robAction(TreeNode root) {
        int[] res = new int[2];

        if (root == null) {
            return res;
        }

        int[] left = robAction(root.left);
        int[] right = robAction(root.right);


        res[0] = Math.max(left[0], left[1]) + Math.max(right[0], right[1]);
        res[1] = root.val + left[0] + right[0];

        return res;
    }
}

File: src/Algorithm/LeetCode/LeetCode_337_House_Robber_III/Solution_DP.py
# Definition for a binary tree node.
from typing import Optional

# https://www.programmercarl.com/0337.打家劫舍III.html
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        res = self.__rob_action(root)

        return max(res[0], res[1])

    def __rob_action(self, root: Optional[TreeNode]):
        if not root:
            return 0, 0

        left = self.__rob_action(root.left)
        right = self.__rob_action(root.right)

        val_0 = max(left[0], left[1]) + max(right[0], right[1])
        val_1 = root.val + left[0] + right[1]

        return val_0, val_1
